### PR TITLE
New Hivelord Ability: Healing Infusion (Replaces Acidic Salve)

### PR DESCRIPTION
**Pull request by Surrealaser**
## About The Pull Request

**1. Healing Infusion (Replaces Acidic Salve)
Plasma Cost:** 200
**Wind Up Time:** 0.5 seconds
**Cooldown:** 5 seconds
**Effect:** Infuses a target xeno within direct line of sight and range 3 with a heal over time for 10 seconds (5 ticks). Whenever the target heals damage normally on a tick, it heals extra damage equal to 10 + 5% of its max HP (this means being on fire can block the effect, and the target must be on weeds). Whenever the target heals Sunder normally on a tick, it heals 3 extra sunder.

Recovery pheromones improve the HP and Sunder regained in this way by 10% per rank.

**2: Minor refactor of Sunder in life.dm:** It now has its own handle proc. Being on fire now disables Sunder recovery. Being on weeds doubles Sunder recovery as does resting. Recovery pheromones improve Sunder recovery by 10% per rank. To compensate, baseline Sunder recovery has been halved.

Ideally I would atomize such a refactor, but I needed a discrete function to call to signal Sunder recovery via Life.dm

**3.** Changed the Hivelord version of Acidic Salve to the Queen's as she still uses it.

## Why It's Good For The Game

Gives the Hivelord a more interesting heal over time option that is distinct and different from other heals with interesting tradeoffs, losing spike, on-demand healing for a considerably more powerful heal over time that can be thwarted by fire and lack of weeds. On that note, it strongly incentivizes Hivelords to spam weeds.

## Changelog
:cl:
add: Healing Infusion ability replaces Acidic Salve for the Hivelord. This is a short-range heal over time that trades spike healing, immunity to fire interruption and independence from weeds for power. Recovery pheromones improve its HP and Sunder recovery by 10% per rank.
refactor: Sunder recovery refactored.
tweak: Baseline Sunder recovery halved.
tweak: Sunder recovery now doubled by weeds and resting, and increased by 10% per rank of recovery pheromones.
tweak: Sunder recovery now blocked by being on fire.
/:cl:

image

### CODE REVIEW

**Comment by neohenry** [comment]
The true support that hivelord is made to be.

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 1]
```diff
@@ -260,8 +260,155 @@ GLOBAL_LIST_INIT(thickenable_resin, typecacheof(list(
...
+	var/heal_range = HIVELORD_HEAL_RANGE
+	var/healing_infusion_aura
+
+
+/datum/action/xeno_action/activable/healing_infusion/can_use_ability(atom/target, silent = FALSE, override_flags)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(QDELETED(target))
+		return FALSE
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:281]: Excess

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 2]
```diff
@@ -260,8 +260,155 @@ GLOBAL_LIST_INIT(thickenable_resin, typecacheof(list(
...
+	plasma_cost = 200
+	keybind_signal = COMSIG_XENOABILITY_HEALING_INFUSION
+	var/heal_range = HIVELORD_HEAL_RANGE
+	var/healing_infusion_aura
+
+
+/datum/action/xeno_action/activable/healing_infusion/can_use_ability(atom/target, silent = FALSE, override_flags)
+	. = ..()
+	if(!.)
+		return FALSE
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:279]: Return

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 3]
```diff
@@ -260,8 +260,155 @@ GLOBAL_LIST_INIT(thickenable_resin, typecacheof(list(
...
+	owner.beam(target,"medbeam",'icons/effects/beam.dmi',10, 10,/obj/effect/ebeam,1)
+	new /obj/effect/temp_visual/telekinesis(get_turf(owner))
+	new /obj/effect/temp_visual/telekinesis(get_turf(target))
+	to_chat(target, "<span class='xenodanger'>Our wounds begin to knit and heal rapidly as [owner]'s healing energies infuse us.</span>") //Let the target know.
+
+	var/mob/living/carbon/xenomorph/patient = target
+
+	patient.has_healing_infusion = TRUE
+	healing_infusion_aura = filter(type = "outline", size = 1, color = COLOR_VERY_PALE_LIME_GREEN) //Set our cool aura; also confirmation we have the buff
+	patient.filters += healing_infusion_aura
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:338]: All these filter related functions should use the correct helpers @Hughgent merge them already pls

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 4]
```diff
@@ -297,6 +297,10 @@
 	var/ignore_pain = FALSE // when true the rav will not go into crit or take crit damage.
 	var/ignore_pain_state = 0 // how far "dead" the rav has got while ignoring pain.
 
+	// *** Hivelord vars *** //
+	///If the target is subject to the Hivelord's Healing Infusion buff this will be true
+	var/has_healing_infusion
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:302]: See filter request

File: code/modules/mob/living/carbon/xenomorph/life.dm [hunk 5]
```diff
@@ -85,6 +82,26 @@
...
+	if(locate(/obj/effect/alien/weeds) in get_turf(src)) //Weeds double sunder recovery
+		sunder_recov *= 2
+
+	if(recovery_aura)
+		sunder_recov *= 1 + recovery_aura * 0.1 //10% bonus per rank of recovery aura
+
+	SEND_SIGNAL(src, COMSIG_XENOMORPH_SUNDER_REGEN, src)
+
+	adjust_sunder(sunder_recov)
+
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/life.dm:104]: Bookmark

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 6]
```diff
@@ -260,8 +260,155 @@ GLOBAL_LIST_INIT(thickenable_resin, typecacheof(list(
...
+	RegisterSignal(patient, COMSIG_XENOMORPH_HEALTH_REGEN, .proc/healing_infusion_regeneration, patient) //Register so we apply the effect whenever the target heals
+	RegisterSignal(patient, COMSIG_XENOMORPH_SUNDER_REGEN, .proc/healing_infusion_sunder_regeneration, patient) //Register so we apply the effect whenever the target heals
+
+
+///Called when the target xeno regains HP via heal_wounds in life.dm
+/datum/action/xeno_action/activable/healing_infusion/proc/healing_infusion_regeneration(datum/source, mob/living/carbon/xenomorph/patient, healing_infusion_filter)
+	SIGNAL_HANDLER
+
+	if(!patient.has_healing_infusion)
+		healing_infusion_deactivate(patient) //if we somehow lose the buff; maybe there's a purge mechanic later, whatever
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:357]: Bookmark

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Review by TiviPlus — COMMENTED** [review]
Reviews didn't submit again

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 7]
```diff
@@ -260,8 +260,148 @@ GLOBAL_LIST_INIT(thickenable_resin, typecacheof(list(
...
+	if(!patient.sunder) //We're at full health with no sunder; no need to proceed
+		return
+
+	new /obj/effect/temp_visual/telekinesis(get_turf(patient)) //Visual confirmation
+
+	patient.adjust_sunder(-3 * (1 + patient.recovery_aura * 0.1)) //-1 + a 10% bonus per rank of our recovery aura
+
+
+///Called when the duration of healing infusion lapses
+/datum/action/xeno_action/activable/healing_infusion/proc/healing_infusion_deactivate(mob/living/carbon/xenomorph/patient)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:394]: Inline this

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 8]
```diff
@@ -260,8 +260,148 @@ GLOBAL_LIST_INIT(thickenable_resin, typecacheof(list(
...
+
+	if(!patient) //No runtimes allowed
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:396]: Unneeded

File: code/modules/mob/living/carbon/xenomorph/life.dm [hunk 9]
```diff
@@ -85,6 +82,26 @@
...
+	if(resting) //Resting doubles sunder recovery
+		sunder_recov *= 2
+
+	if(locate(/obj/effect/alien/weeds) in get_turf(src)) //Weeds double sunder recovery
+		sunder_recov *= 2
+
+	if(recovery_aura)
+		sunder_recov *= 1 + recovery_aura * 0.1 //10% bonus per rank of recovery aura
+
+	SEND_SIGNAL(src, COMSIG_XENOMORPH_SUNDER_REGEN, src)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/life.dm:101]: No need to send src as an arg, signals already do that

File: code/modules/mob/living/carbon/xenomorph/life.dm [hunk 10]
```diff
@@ -106,6 +123,9 @@
 			regen_power = min(regen_power + xeno_caste.regen_ramp_amount*20,1)
 		amount *= regen_power
 	amount *= multiplier
+
+	SEND_SIGNAL(src, COMSIG_XENOMORPH_HEALTH_REGEN, src)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/life.dm:127]: Ditto

File: code/modules/mob/living/carbon/xenomorph/life.dm [hunk 11]
```diff
@@ -85,6 +82,26 @@
...
+	if(!sunder || on_fire) //No sunder, no problem; or we're on fire and can't regenerate.
+		return
+
+	var/sunder_recov = xeno_caste.sunder_recover * -0.5 //Baseline
+
+	if(resting) //Resting doubles sunder recovery
+		sunder_recov *= 2
+
+	if(locate(/obj/effect/alien/weeds) in get_turf(src)) //Weeds double sunder recovery
+		sunder_recov *= 2
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/life.dm:96]:
```suggestion
	if(locate(/obj/effect/alien/weeds) in loc) //Weeds double sunder recovery
		sunder_recov *= 2
```

**Review by TiviPlus — APPROVED** [review]
last one

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 12]
```diff
@@ -1,3 +1,11 @@
+//Some debug variables. Toggle them to 1 in order to see the related debug messages. Helpful when testing out formulas.
+#define DEBUG_HIVELORD_ABILITIES	0
+#if DEBUG_HIVELORD_ABILITIES
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:3]: Ifdef

File: code/modules/mob/living/carbon/xenomorph/life.dm [hunk 13]
```diff
@@ -85,6 +82,26 @@
 			heal_wounds(XENO_STANDING_HEAL * ruler_healing_penalty, TRUE) //Major healing nerf if standing.
 	updatehealth()
 
+/mob/living/carbon/xenomorph/proc/handle_living_sunder_updates()
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/life.dm:85]: Autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm (hunk 12 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:3]: This actually doesn't work unfortunately; I tried it and had to revert, unless you're telling me to ifdef all of the debug message instances which is a less prefable solution.

Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:3]: No I mean just
//define DEBUG_HIVELORD_ABILITIES "ae"
#ifdef DEBUG_HIVELORD_ABILITIES

Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:3]: pending still

**Review by TiviPlus — APPROVED** [review]

**Comment by Surrealaser** [comment]
Can we get a TM on this?

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**